frontend/ui/widget/notification.lua

```diff
@@ -48,44 +53,73 @@ function Notification:init()
         end
     end
 
-    -- we construct the actual content here because self.text is only available now
     local text_widget = TextWidget:new{
         text = self.text,
         face = self.face,
     }
     local widget_size = text_widget:getSize()
-    self[1] = CenterContainer:new{
-        dimen = Geom:new{
-            w = Screen:getWidth(),
-            h = math.floor(Screen:getHeight() / 10),
-        },
-        FrameContainer:new{
-            background = Blitbuffer.COLOR_WHITE,
-            radius = 0,
-            margin = self.margin,
-            padding = self.padding,
-            CenterContainer:new{
-                dimen = Geom:new{
-                    w = widget_size.w,
-                    h = widget_size.h
-                },
-                text_widget,
-            }
+    self.frame = FrameContainer:new{
+        background = Blitbuffer.COLOR_WHITE,
+        radius = 0,
+        margin = self.margin,
+        padding = self.padding,
+        CenterContainer:new{
+            dimen = Geom:new{
+                w = widget_size.w,
+                h = widget_size.h
+            },
+            text_widget,
         }
     }
+    local notif_height = self.frame:getSize().h
+
+    self:_cleanShownStack()
+    table.insert(Notification._nums_shown, os.time())
+    self.num = #Notification._nums_shown
+
+    self[1] = VerticalGroup:new{
+        align = "center",
+        -- We use Spans to properly position this notification:
+        -- have this VerticalGroup full width, to ensure centering
+        HorizontalSpan:new{ width = Screen:getWidth() },
+        -- push this frame at its y=self.num position
+        VerticalSpan:new{ width = notif_height * (self.num - 1) },
+        self.frame,
+    }
+end
+
+function Notification:_cleanShownStack(num)
+    -- Clean stack of shown notifications
+    if num then
+        -- This notification is no longer displayed
+        Notification._nums_shown[num] = false
+    end
+    -- Remove from the stack all slots no longer displayed
+    -- from the end: new notifications can be shown there.
+    -- As a sanity check, we also forget those shown for
+    -- more than 30s in case no close event was received.
+    local expire_ts = os.time() - 30
+    for i=#Notification._nums_shown, 1, -1 do
+        if Notification._nums_shown[i] and Notification._nums_shown[i] > expire_ts then
+            break -- still shown (or not yet expired)
+        end
+        table.remove(Notification._nums_shown, i)
+    end
```

Comment:
Won't this fail to remove a non-tail index, if the tail itself is still shown and not expired?

(e.g., I would expect a `continue` instead of a `break` given the comment. If you forget the fact that Lua doesn't have a `continue` keyword ;p).

Reply:
My idea was to only remove tails. Even if slots have been freed at top, it felt nicer to add new one at bottom. Easier for the eyes to follow what is happening.
I can add that ^ as a comment if you agree with the idea.

Reply:
Right, that's because it's used for the vertical UI alignment, I hadn't made the connection ;).

Yep, makes perfect sense then, so, yeah, I wouldn't mind a better comment ;).